File: src/main/java/ru/arc/eliteloot/DecorPool.java
package ru.arc.eliteloot;

import java.util.TreeMap;
import java.util.concurrent.ThreadLocalRandom;

import lombok.Getter;
import lombok.ToString;

import static ru.arc.util.Logging.debug;

@Getter
@ToString
public class DecorPool {

    TreeMap<Double, DecorItem> decors = new TreeMap<>();

    public void add(DecorItem decorItem, double weight) {
        debug("Adding decor item: {} with weight: {}", decorItem.getMaterial(), weight);
        if (weight == 0) return;
        double lastKey = decors.isEmpty() ? 0 : decors.lastKey();
        decors.put(lastKey + weight, decorItem);
    }

    public DecorItem randomItem() {
        //ARC.info(decors.toString());
        if (decors.isEmpty()) return null;
        double random = ThreadLocalRandom.current().nextDouble(0, decors.lastKey());
        return decors.ceilingEntry(random).getValue();
    }

    public boolean contains(DecorItem decorItem) {
        return decors.containsValue(decorItem);
    }

    public void remove(DecorItem decorItem) {
        decors.values().removeIf(decorItem::equals);
    }
}

File: src/main/kotlin/ru/arc/eliteloot/DecorPool.kt
package ru.arc.eliteloot

import ru.arc.util.Logging.debug
import java.util.TreeMap
import java.util.concurrent.ThreadLocalRandom

class DecorPool {

    val decors: TreeMap<Double, DecorItem> = TreeMap()

    fun add(decorItem: DecorItem, weight: Double) {
        debug("Adding decor item: {} with weight: {}", decorItem.material, weight)
        if (weight == 0.0) return
        val lastKey = if (decors.isEmpty()) 0.0 else decors.lastKey()
        decors[lastKey + weight] = decorItem
    }

    fun randomItem(): DecorItem? {
        if (decors.isEmpty()) return null
        val random = ThreadLocalRandom.current().nextDouble(0.0, decors.lastKey())
        return decors.ceilingEntry(random)?.value
    }

    fun contains(decorItem: DecorItem): Boolean = decors.containsValue(decorItem)

    fun remove(decorItem: DecorItem) {
        decors.values.removeIf { it == decorItem }
    }

    override fun toString(): String = "DecorPool(decors=$decors)"
}
